src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitUsingTests.cs

```diff
@@ -1636,5 +1651,404 @@ public class D
+        }
+
+        System.Console.Write(""return"");
+        return 1;
+    }
+    public async System.Threading.Tasks.ValueTask DisposeAsync(int i = 0)
+    {
+        System.Console.Write($""dispose_start "");
+        await System.Threading.Tasks.Task.Yield();
+        System.Console.Write($""dispose_end "");
+    }
+}
+";
+            var comp = CreateCompilationWithTasksExtensions(new[] { source, s_interfaces }, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "using dispose_start dispose_end return");
+        }
+
+        [ConditionalFact(typeof(WindowsDesktopOnly))]
+        [WorkItem(32316, "https://github.com/dotnet/roslyn/issues/32316")]
+        public void TestPatternBasedDisposal_InstanceMethod_ParamsParameter()
+        {
+            string source = @"
+public class C
+{
+    public static async System.Threading.Tasks.Task<int> Main()
+    {
+        await using (var x = new C())
+        {
+            System.Console.Write(""using "");
+        }
+
+        System.Console.Write(""return"");
+        return 1;
+    }
+    public async System.Threading.Tasks.ValueTask DisposeAsync(params int[] x)
+    {
+        System.Console.Write($""dispose_start "");
+        await System.Threading.Tasks.Task.Yield();
+        System.Console.Write($""dispose_end "");
+    }
+}
+";
+            var comp = CreateCompilationWithTasksExtensions(new[] { source, s_interfaces }, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "using dispose_start dispose_end return");
+        }
+
+        [ConditionalFact(typeof(WindowsDesktopOnly))]
+        [WorkItem(32316, "https://github.com/dotnet/roslyn/issues/32316")]
+        public void TestPatternBasedDisposal_InstanceMethod_ReturningVoid()
+        {
+            string source = @"
+public class C
+{
+    public static async System.Threading.Tasks.Task<int> Main()
+    {
+        await using (var x = new C())
+        {
+        }
+
+        return 1;
+    }
+    public void DisposeAsync()
+    {
+    }
+}
+";
+            var comp = CreateCompilationWithTasksExtensions(new[] { source, s_interfaces });
+            comp.VerifyDiagnostics(
+                // (6,22): error CS4008: Cannot await 'void'
+                //         await using (var x = new C())
+                Diagnostic(ErrorCode.ERR_BadAwaitArgVoidCall, "var x = new C()").WithLocation(6, 22)
+                );
+        }
+
+        [ConditionalFact(typeof(WindowsDesktopOnly))]
+        [WorkItem(32316, "https://github.com/dotnet/roslyn/issues/32316")]
+        public void TestPatternBasedDisposal_InstanceMethod_ReturningInt()
+        {
+            string source = @"
+public class C
+{
+    public static async System.Threading.Tasks.Task<int> Main()
+    {
+        await using (var x = new C())
+        {
+        }
+
+        return 1;
+    }
+    public int DisposeAsync()
```

Comment:
The latter is a great case. Currently it is accepted, because you can indeed do `await enumerator.DisposeAsync();`.
Do you think it should be rejected?